### DISCUSSION TITLE
How to make ImageSharp.Web intercept all image requests including these without querystring

### DISCUSSION THREAD

**Discussion by RosenPetrov** [SixLabors/ImageSharp.Web#315] [Q&A] [answered]
Hey,

I am working on upgrading a solution which was using ImageProcessor.Wev to ImageSharp.Web. I have a requirement to remove the metadata of all images on my website. I found how to remove the metadata, but I want also to handle the images which URLs don't contain query string. In ImageProcessorWeb package there was a setting called interceptAllRequests (https://imageprocessor.org/imageprocessor-web/configuration/) How can I accomplish this behavior now with ImageSharp.Web? 

Thanks

**Comment by JimBobSquarePants** [comment]
It's a bit different with ImageSharp.Web

Whether all requests should be captured it the responsibility of the individual [IImageProvider](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.Providers.IImageProvider.html#SixLabors_ImageSharp_Web_Providers_IImageProvider_ProcessingBehavior) implementations.

For the `PhysicalFileSystemProvider` you can configure this behavior on [startup](https://docs.sixlabors.com/articles/imagesharp.web/gettingstarted.html#setup-and-configuration).

```c#
services.AddImageSharp()
    .Configure<PhysicalFileSystemProviderOptions>(options =>
    {
        options.ProcessingBehavior = ProcessingBehavior.All;
    });
```

**Reply by RosenPetrov** [reply]
Hello, 

Thanks for the reply. It didn't work for me straight away. My project is Umbraco based. I checked in their source code and I saw that they actually do use WebRootImageProvider and I couldn't find a way to update the processing behavior because it looks like there are no sutiable configuration options for WebRootImageProvider. 

I ended up creating a custom WebRootImageProvider where I just set the ProcessingBehavior to All. Here is the code
```
public sealed class CustomWebRootImageProvider : FileProviderImageProvider
{
    public WebRootImageProvider(
        IWebHostEnvironment environment,
        FormatUtilities formatUtilities)
        : base(environment.WebRootFileProvider, ProcessingBehavior.All, formatUtilities)
    { }
}
```

Then to ensure my Provider is used I had to do this in my Umbraco Composer (Umbraco are calling AddImageSharp() already I don't have access to the IImageSharpBuilder and I didn't want to call AddImageSharp for a second time in the startup)

```
builder.Services.RemoveAll<IImageProvider>();
builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IImageProvider, CustomWebRootImageProvider>()); 
```

This worked, and I can see that now images without query string are processed. But what I can't get is why my code for removing the metadata is not applied on the images without query string? Do you have any idea? This is the configuration I have in place about that

```
public sealed class ConfigureImageSharpMiddlewareOptions : IConfigureOptions<ImageSharpMiddlewareOptions>
{
    public void Configure(ImageSharpMiddlewareOptions options)
    {
        options.OnBeforeSaveAsync = f =>
        {
            f.Image.Metadata.ExifProfile = null;
            f.Image.Metadata.XmpProfile = null;
            f.Image.Metadata.IptcProfile = null;
            return Task.CompletedTask;
        };
    }
}
```

**Comment by JimBobSquarePants** [comment]
You likely want to use post configuration. I’ve no idea what the Umbraco configuration pipeline is. It’s very custom 

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-7.0#options-post-configuration

**Comment by RosenPetrov** [comment, marked as answer]
Thanks for your reply again! Excuse me for bothering with Umbraco which is not relevant for this ticket. 

I found out why the metadata is not removed for images with no query string. In ImageSharpMiddleware the "commands" are parsed from the query string and even if there are some checks which let the processing happen when ProcessingBehavior.All almost at the end there is this check

```
// No commands? We simply copy the stream across.
if (commands.Count == 0)
{
    await inStream.CopyToAsync(outStream);
    outStream.Position = 0;
    format = await Image.DetectFormatAsync(this.options.Configuration, outStream);
}
else
{
    FormattedImage image = null;
    try
    {
        // Now we can finally process the image.
        // We first sort the processor collection by command order then use that collection to determine whether the decoded image pixel format
        // explicitly requires an alpha component in order to allow correct processing.
        //
        // The non-generic variant will decode to the correct pixel format based upon the encoded image metadata which can yield
        // massive memory savings.
        IReadOnlyList<(int Index, IImageWebProcessor Processor)> sortedProcessors = this.processors.OrderBySupportedCommands(commands);
        bool requiresAlpha = sortedProcessors.RequiresTrueColorPixelFormat(commands, this.commandParser, this.parserCulture);

        if (requiresAlpha)
        {
            image = await FormattedImage.LoadAsync<Rgba32>(this.options.Configuration, inStream);
        }
        else
        {
            image = await FormattedImage.LoadAsync(this.options.Configuration, inStream);
        }

        image.Process(
            this.logger,
            sortedProcessors,
            commands,
            this.commandParser,
            this.parserCulture);

        await this.options.OnBeforeSaveAsync.Invoke(image);

        image.Save(outStream);
        format = image.Format;
    }
    finally
    {
        image?.Dispose();
    }
}
```

I end up in the if statement and the OnBeforeSaveAsync is not executed, so the metadata is not removed. I am wondering now f you have a suggestion on how to work around that. Now I remember the for ImageProcessor.Web I had included the package source code and modified the logic, but I would like not to do that here.

**Reply by JimBobSquarePants** [reply]
Ah! I better fix that for V3! Can you please raise an issue so I remember?

As a workaround I would add a noop [`IImageWebProcessor`](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.Processors.IImageWebProcessor.html) and invisibly add a command for it via [`ImageSharpMiddlewareOptions.OnParseCommandsAsync` ](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddlewareOptions.html#SixLabors_ImageSharp_Web_Middleware_ImageSharpMiddlewareOptions_OnParseCommandsAsync)

**Reply by RosenPetrov** [reply]
Thanks! I was in the right direction. Here is my noop IImageProcessor:

```
 public class RemoveMetadataProcessor : IImageWebProcessor
  {
      public IEnumerable<string> Commands => new[] { "removeMetadata" };

      public FormattedImage Process(FormattedImage image, ILogger logger, CommandCollection commands, CommandParser parser, CultureInfo culture)
      {
          return image;
      }

      public bool RequiresTrueColorPixelFormat(CommandCollection commands, CommandParser parser, CultureInfo culture) => false;
  }
```

And I was just about to replace the IRequestParser to include the query string when I saw your comment, so this is the OnParseCommandsAsync code:

```
 options.OnParseCommandsAsync = f =>
  {
      f.Commands.Add("removeMetadata", "true");
      return Task.CompletedTask;
  };
```

And if somebody is curios how I added the processor, this is the line in my Umbraco Composer

`  builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IImageWebProcessor, RemoveMetadataProcessor>());`

Works this way. Have a nice day!

**Comment by JimBobSquarePants** [comment]
I just had a thought…
You don’t need the OnBeforeSave event if you remove the metadata in your processor.
